activesupport/lib/active_support/message_verifier.rb

```diff
@@ -64,54 +66,162 @@ def valid_message?(signed_message)
+    # given in the options.
+    #
+    #   verifier.verified(signed_message, for: 'something_else') # raises ActiveSupport::Claims::InvalidClaims
+    #
+    # Raises +ExpiredClaims+ if the message is expired.
+    #
+    #   expired_message = verifier.generate('a private message', expires_at: 1.day.ago)
+    #   verifier.verified(signed_message) # raises ActiveSupport::Claims::ExpiredClaims
+    def verified(signed_message, options = {})
+      if signed_message.include? '--'
+        legacy_verifier = LegacyMessageVerifier.new(@secret, digest: @digest, serializer: @serializer)
+        legacy_verifier.verify(signed_message)
+      else
+        return unless valid_message?(signed_message)
+
+        data = signed_message.split('.')[1]
+        claims = @serializer.load(decode(data)).symbolize_keys
+        Claims.verify!(claims, options)
       end
+    rescue ArgumentError => argument_error
+      raise unless argument_error.message =~ %r{invalid base64}
     end
 
-    # Decodes the signed message using the +MessageVerifier+'s secret.
+    # Decodes the signed message using the +MessageVerifier+'s secret. It also decodes
+    # legacy signed messages which are outdated messages generated by +MessageVerifier+
+    # of the previous rails version.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
-    #   signed_message = verifier.generate 'a private message'
+    #   signed_message = verifier.generate('a private message', for: 'test', expires_in: 1.day)
+    #
+    #   # Within 1 day...
+    #   verifier.verify(signed_message, for: 'test') # => "a private message"
     #
-    #   verifier.verify(signed_message) # => 'a private message'
+    #   legacy_signed_message = "BAhJIhZhIHByaXZhdGUgbWVzc2FnZQY6BkVU--a2a1232deaf49ba164ed0d0b43b3f17cf8026c46"
+    #   verifier.verify(legacy_signed_message) # => "a private message"
     #
     # Raises +InvalidSignature+ if the message was not signed with the same
     # secret or was not Base64-encoded.
     #
     #   other_verifier = ActiveSupport::MessageVerifier.new 'd1ff3r3nt-s3Krit'
-    #   other_verifier.verify(signed_message) # => ActiveSupport::MessageVerifier::InvalidSignature
-    def verify(signed_message)
-      verified(signed_message) || raise(InvalidSignature)
+    #   other_verifier.verify(signed_message) # raises ActiveSupport::MessageVerifier::InvalidSignature
+    def verify(signed_message, options = {})
+      verified(signed_message, options) || raise(InvalidSignature)
     end
 
-    # Generates a signed message for the provided value.
+    # Generates a signed message for the given +value+ and +options+ hash.
     #
     # The message is signed with the +MessageVerifier+'s secret. Without knowing
-    # the secret, the original value cannot be extracted from the message.
+    # the secret, the original value cannot be extracted from the message. A
+    # message is signed in the form of JSON Web Token(JWT). The final message is
```

Comment:
`signed as a...`

Also put a space before the abbreviation parentheses (here and all other places): `JSON Web Token (JWT)`.